File: Remove Nodes From Linked List.cpp

struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
public:
    ListNode* removeNodes(ListNode* head) {
        ListNode *dummy,*temp;
        dummy=new ListNode();
        temp=dummy;

        while(head){
            int val=head->val;
            ListNode *nextt=head->next;
            int flag=0;
            while(nextt){
                if(nextt->val>val){
                    flag=1;
                    break;
                }
                nextt=nextt->next;
            }
            if(flag==0){
                temp->next=head;
                temp=temp->next;
            }
            head=head->next;
        }
        temp->next=head;
        return dummy->next;
    }
   
    
};

File: 2573-remove-nodes-from-linked-list/remove-nodes-from-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeNodes(self, head: Optional[ListNode]) -> Optional[ListNode]:
        #just iterating in reverse and solving the same logic 
        def reverse(root):
            prev=None
            cur=root
            while cur:
                frwd=cur.next
                cur.next=prev
                prev=cur
                cur=frwd
            return prev
        
        reverse_head=reverse(head)
        head=reverse_head
        max_val=head.val
        while head.next:
            if head.next.val<max_val:
                head.next=head.next.next
            else:
                max_val=head.next.val
                head=head.next
        return reverse(reverse_head)
        